### PR TITLE
[CARBONDATA-3528] refactor java checkstyle rules

### PR DESCRIPTION
**Pull request by lamber-ken**
## Checks
Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed? NO
 
 - [x] Any backward compatibility impacted? NO
 
 - [x] Document update required? NO

 - [x] Testing done YES
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. Know

## Detail

**AIMS**
Goog products require good coding checkstyle. This issue aims to refactor the java checkstyle.

**Docs and tools**
* https://checkstyle.sourceforge.io/checks.html
* mvn org.apache.maven.plugins:maven-checkstyle-plugin:2.17:check 

**New rules**
* RedundantImport
* AvoidStarImport
* LeftCurly
* SimplifyBooleanExpression
* EmptyStatement
* EmptyLineSeparator
* WhitespaceAfter
* NoWhitespaceAfter
* NoWhitespaceBefore
* OperatorWrap

**Tasks**
Current, the severity property of these new rules is `info` level, so it will not affect the build. 
After we finish these rules, the level will change to `error` level.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/587/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/590/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/592/

**Comment by lamber-ken** [comment]
Hi @jackylk, if you have time, can you please take a look?

**Comment by ajantha-bhat** [comment]
@lamber-ken : 
Please update the description and I have some questions ?
1. Why this change now ? what is the problem with current check style ? these new rules are taken from which reference or project ?
2. Why only java checkstyle modification ? why not scala check style also ?
3. After adding these rules ? no issues found in carbon ?

File: dev/javastyle-config.xml [hunk 1]
```diff
@@ -1,173 +1,270 @@
...
+-->
 
 <!DOCTYPE module PUBLIC
           "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
           "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
 
 <!--
 
-    Checkstyle configuration based on the Google coding conventions from:
+  Checkstyle configuration based on the Google coding conventions from:
```
Comment (ajantha-bhat) [inline dev/javastyle-config.xml]: Also keep only new changes, please revert other changes

Comment (lamber-ken) [reply dev/javastyle-config.xml]: see

**Comment by lamber-ken** [comment]
> @lamber-ken :
> Please update the description and I have some questions ?
> 
> 1. Why this change now ? what is the problem with current check style ? these new rules are taken from which reference or project ?
> 2. Why only java checkstyle modification ? why not scala check style also ?
> 3. After adding these rules ? no issues found in carbon ?

hi, @ajantha-bhat,

This issus just fucus on java checkstyle, it's a big work to refactor java checkstyle already, scala checkstyle needs a new thread to refactor. And some new rules, for example bellow.

you can check rulues by `mvn org.apache.maven.plugins:maven-checkstyle-plugin:2.17:check`.

#### RedundantImport (org.apache.carbondata.core.scan.expression.MatchExpression)
```
import org.apache.carbondata.common.annotations.InterfaceAudience;
import org.apache.carbondata.core.metadata.datatype.DataTypes;
import org.apache.carbondata.core.scan.expression.Expression;
import org.apache.carbondata.core.scan.expression.ExpressionResult;
import org.apache.carbondata.core.scan.expression.exception.FilterIllegalMemberException;
import org.apache.carbondata.core.scan.expression.exception.FilterUnsupportedException;
import org.apache.carbondata.core.scan.filter.intf.ExpressionType;
import org.apache.carbondata.core.scan.filter.intf.RowIntf;
```

#### @Override (org.apache.carbondata.datamap.PreAggregateDataMapProvider)
```
@Override
public boolean supportRebuild() {
  return false;
}

@Override public boolean rebuildInternal(String newLoadName, Map<String, List<String>> segmentMap,
    CarbonTable carbonTable) {
  return false;
}
```

#### AvoidStarImport (org.apache.carbondata.tool.DataSummary)
```
import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.*;
```

#### EmptyStatement (org.apache.carbondata.datamap.lucene.LuceneDataMapFactoryBase)
```
// validate DataMapSchema and get index columns
indexedCarbonColumns =  carbonTable.getIndexedColumns(dataMapSchema);;
flushCacheSize = validateAndGetWriteCacheSize(dataMapSchema);
storeBlockletWise = validateAndGetStoreBlockletWise(dataMapSchema);
```
and so on.

File: dev/javastyle-config.xml (hunk 1 above)
Comment (lamber-ken) [reply dev/javastyle-config.xml]: done.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/668/

File: dev/javastyle-config.xml [hunk 2]
```diff
@@ -42,132 +42,229 @@
...
-        </module>
-        <module name="EmptyCatchBlock">
-            <property name="exceptionVariableName" value="expected"/>
-        </module>
-        <module name="CommentsIndentation"/>
-        <module name="UnusedImports"/>
-
-        <module name="ParenPad"/>
-        <module name="WhitespaceAround"/>
+  <property name="charset" value="UTF-8"/>
```
Comment (ajantha-bhat) [inline dev/javastyle-config.xml]: can you add the new properties at the end ? cannot differentiate old and new. Now it shows both are modified. 
it should show only modified/added properties. 
Existing properties should not be deleted

**Comment by ajantha-bhat** [comment]
How many new error(info, as now the level is set to info) comes after adding this checkstyle ? 

If we are not fixing those errors why are we adding the rules ?

@ravipesala , @kumarvishal09 please check

File: dev/javastyle-config.xml (hunk 2 above)
Comment (lamber-ken) [reply dev/javastyle-config.xml]: ok, I see. Done.

**Comment by lamber-ken** [comment]
> How many new error(info, as now the level is set to info) comes after adding this checkstyle ?
> 
> If we are not fixing those errors why are we adding the rules ?
> 
> @ravipesala , @kumarvishal09 please check

Hi, @ajantha-bhat, I'll fix these style errors at the futuer.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/670/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/674/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/674/

**Comment by ajantha-bhat** [comment]
Changes looks ok after the comments now.
@kumarvishal09 , @jackylk please conclude do we need these new java checkstyle now.

**Comment by lamber-ken** [comment]
> @ajantha-bhat
> I think these new rules are good for improving code readability.
> After introducing these rules which is INFO level, we can resolve it in the future step by step.
> So, suggest to merge it. What do you think?

thanks, I'll  resolve it in the future step by step.

**Comment by ajantha-bhat** [comment]
LGTM

**Comment by lamber-ken** [comment]
@jackylk, please cc, thanks.

**Comment by zzcclp** [comment]
Will merge this later. Thanks @lamber-ken

**Pull request closed without merging**